### DISCUSSION TITLE
Calculating stiffness matrix for specific cells in mesh

### DISCUSSION THREAD

**Discussion by jmm1828** [firedrakeproject/firedrake#3031] [Firedrake support] [answered]
I have a simple 2D mesh, say `mesh=UnitSquareMesh(2,2)`, and I want to calculate the stiffness matrix for the bilinear form a(u,v) in my linear variational problem at different cells in the mesh. I can calculate a "global" stiffness matrix as follows:

```
# create mesh
m = UnitSquareMesh(2,2)
# Create piecewise linear function space
V=FunctionSpace(m,"CG",1)
# Create test and trial functions
v=TestFunction(V)
u=TrialFunction(V)
# create bilinear form a(u,v) = ∫ uv dx
a = u*v*dx
# calculate stiffness matrix
assemble(a)
...
```

But is there a way to perform this calculation cell by cell? Any help would be much appreciated.

**Comment by wence-** [comment]
If you make a `BrokenElement` version of your finite elements then it will have discontinuous (cell-wise) continuity and that will allow you to get cell-wise matrices. Although it won't work if you have facet integrals I think.

But, what do you want to _do_ with these cell-wise matrices? There might well be a better way to achieve what you want.

**Reply by jmm1828** [reply]
I want to modify my variational problem at different cells in the mesh, depending on certain characteristics of each cell, and then solve these modified equations on each cell.

**Reply by dham** [reply]
What sort of modification do you want to make on each cell? There are various ways this could be done (conditionals, indicator functions...) depending exactly on what sort of change you want to make and what characteristics of the cell you are using. It would help if you could show us the maths you are trying to represent.

**Reply by jmm1828** [reply]
On each cell I want to add the term $\int_\Omega D_k(\nabla u_k \cdot \nabla v_k)  \ dx$ to the weak formulation of by problem, for stability purposes. 
Where  $u_k$ and $v_k$ are trial and test functions on the cell respectively and $D_k$ is a cell dependant matrix.
Specifically, $D_k$ is calculated by taking normalised tangent vectors along the edges of each cell, calculating outer products of these vectors with themselves and adding the resulting products together. I already have $D_k$ for each cell.

**Reply by wence-** [reply]
Do you have $D_k$ as a `Function`? If so, then you can just add `D_k*inner(grad(u), grad(v))*dx` to your bilinear form. Although this seems like it is slightly wrong because if $D_k \in \mathbb{R}^{2 \times 2}$ say then the integrand as your wrote it isn't scalar?

**Reply by wence-** [reply]
If I recall correctly, we determined how to make this `D_k` by interpolating an expression into a DG0 tensor function space. You can probably also avoid that, if you have the symbolic expression for `D_k` (call it `D_k_expr`), then you can avoid the interpolation and just write:
```
a_stabilised = a + D_k_expr*inner(grad(u), grad(v))*dx
```

**Reply by jmm1828** [reply]
My apologies, the integrand should be a scalar. This is the correct way to write it $\int_\Omega( D_k\nabla u_k) \cdot \nabla v_k  \ dx$

**Reply by jmm1828** [reply]
I will try this, thank you.

**Reply by jmm1828** [reply]
If I have a symbolic expressions the $D_K$ s  how can I write them as a function of the cells?

**Reply by wence-** [reply]
They are automatically so (in the same way that `TestFunction(V)` is a function of the cells). I think you had used `CellEdgeVectors` and so you have:
```python
cev = CellEdgeVectors(mesh) # symbolic expression that is per-cell when assembled/interpolated
D_k = outer(cev[0, :], cev[0, :]) + outer(cev[1, :], cev[1, :]) + outer(cev[2, :], cev[2, :]) # for triangles

a_stabilisation = inner(dot(D_k, grad(u_k)), grad(v_k))*dx

a_stabilised = a + a_stabilisation
```

(untested)

**Reply by jmm1828** [reply]
Thank you for this, and would it be possible to assign weights to different edges with this method?
Say for a given cell each edge has a weight and once you calculate an outer product you multiply the matrix by this weight before adding.

**Reply by wence-** [reply]
This is a little trickier since you need to be able to associate the edge weights with the appropriate cell edge vector. How are you generating and storing the edge weights at the moment?

**Reply by jmm1828** [reply]
I have not generated the weights yet, they would depend on the length of the edge and the shape regularity of the mesh. I wanted to try some arbitrary weights first

**Reply by wence-** [reply, marked as answer]
Arbitrary weights are in some sense harder than depending on length and shape regularity since you need to be able to associate them with the correct edges and then pull them out again (which is not trivial because if you store them as a Function they will be evaluated at quadrature points whereas you want to "see" the basis vector).

On the other hand, the length of each relevant edge can be computed with `sqrt(dot(cev[0, :], cev[0, :]))` (for edge 0, say). I do not think there is a symbolic expression for the shape regularity coefficient of a given cell, however, this is easier. If you have some way of computing it, you can store it in a `shape_regularity = Function(FunctionSpace(mesh, "DG", 0))` and then use that in your form.

**Reply by wence-** [reply]
I guess if you're using 2D elements and you want the condition number of the Jacobian as your regularity metric then you can use that for a 2x2 matrix $J$ that the eigenvalues are:

$$
\text{tr}(J) / 2 \pm \sqrt{\big(\text{tr}(J) / 2\big)^2 - |J|}
$$

So

```python
J = Jacobian(mesh)
detJ = JacobianDeterminant(mesh)
m = tr(J) / 2

kappa = (m + sqrt(m**2 - p))**2 / p

```

Edit: if you go this route, you'll probably want to manually select an appropriate quadrature degree for the stabilisation term (`...*dx(degree=degree_of_polynomial_you_want_to_integrate_exactly)`) since UFL will otherwise estimate something that is too high.

**Reply by jmm1828** [reply]
Thank you for this. I just wanted arbitrary weights for testing purposes, but this is probably a better route.
